dataplane/standalone/saiserver/acl.go

```diff
@@ -0,0 +1,167 @@
+	tableToBank map[uint64]groupBank
+	// groupNextFreeBank contains the next free bank for a group.
+	groupNextFreeBank map[uint64]int
+}
+
+func newACL(mgr *attrmgr.AttrMgr, dataplane aclDataplaneAPI, s *grpc.Server) *acl {
+	a := &acl{
+		mgr:       mgr,
+		dataplane: dataplane,
+	}
+	saipb.RegisterAclServer(s, a)
+	return a
+}
+
+// CreateAclTableGroup creates a lucius flow table, where the group members are banks in the flow table.
+func (a *acl) CreateAclTableGroup(ctx context.Context, req *saipb.CreateAclTableGroupRequest) (*saipb.CreateAclTableGroupResponse, error) {
+	id := a.mgr.NextID()
+
+	stage := req.GetAclStage()
+	typ := req.GetType()
+	bind := req.GetAclBindPointTypeList()
+	if stage != saipb.AclStage_ACL_STAGE_EGRESS && stage != saipb.AclStage_ACL_STAGE_PRE_INGRESS && stage != saipb.AclStage_ACL_STAGE_INGRESS {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid stage type: %v", stage)
+	}
+	if typ != saipb.AclTableGroupType_ACL_TABLE_GROUP_TYPE_PARALLEL {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid type: %v", typ)
+	}
+	if len(bind) != 1 && bind[0] == saipb.AclBindPointType_ACL_BIND_POINT_TYPE_SWITCH {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid bind configuration: %v", typ)
+	}
+
+	tReq := &fwdpb.TableCreateRequest{
+		ContextId: &fwdpb.ContextId{Id: a.dataplane.ID()},
+		Desc: &fwdpb.TableDesc{
+			TableType: fwdpb.TableType_TABLE_TYPE_FLOW,
+			TableId: &fwdpb.TableId{
+				ObjectId: &fwdpb.ObjectId{
+					Id: fmt.Sprint(id),
+				},
+			},
+			Table: &fwdpb.TableDesc_Flow{
+				Flow: &fwdpb.FlowTableDesc{
+					BankCount: 1,
+				},
+			},
+		},
+	}
+	if _, err := a.dataplane.TableCreate(ctx, tReq); err != nil {
+		return nil, err
+	}
+
+	return &saipb.CreateAclTableGroupResponse{Oid: 1}, nil
+}
+
+// CreateAclTableGroupMember stores the acl table id to its corresponding lucius flow table and bank.
+func (a *acl) CreateAclTableGroupMember(_ context.Context, req *saipb.CreateAclTableGroupMemberRequest) (*saipb.CreateAclTableGroupMemberResponse, error) {
+	groupID := req.GetAclTableGroupId()
+	tableID := req.GetAclTableId()
+	memberID := a.mgr.NextID()
+
+	bank := a.groupNextFreeBank[groupID]
+	a.groupNextFreeBank[groupID] = bank + 1
+	a.tableToBank[tableID] = groupBank{
+		groupID:  fmt.Sprint(groupID),
+		bank:     bank,
+		memberID: memberID,
+	}
+
+	return &saipb.CreateAclTableGroupMemberResponse{Oid: memberID}, nil
+}
+
+// CreateAclTable is noop as the table is already created in the group.
+// TODO: Do we need to support tables that aren't in groups.
+func (a *acl) CreateAclTable(_ context.Context, req *saipb.CreateAclTableRequest) (*saipb.CreateAclTableResponse, error) {
+	id := a.mgr.NextID()
+	return &saipb.CreateAclTableResponse{Oid: id}, nil
+}
+
+// CreateAclEntry adds an entry in the a bank.
+func (a *acl) CreateAclEntry(ctx context.Context, req *saipb.CreateAclEntryRequest) (*saipb.CreateAclEntryResponse, error) {
+	id := a.mgr.NextID()
```

Comment:
So NextID() is also called here in addition to CreateAclTableGroupMember, is that right? I'm not really understanding the code but wanted to point that out.

Reply:
yes, every object created needs an id